### ISSUE TITLE
HDR colors look washed out while playing Native HDR games (or even via Moonlight/Chiaki HDR)

### ISSUE THREAD

**Issue by arshiatn** [ValveSoftware/SteamOS#1314]
#### Your system information

* Steam client version: 1702079146
* SteamOS version: 3.5.7
* Opted into Steam client beta?: No
* Opted into SteamOS beta?: No 
* Have you checked for updates in Settings > System?: Yes

#### Please describe your issue in as much detail as possible:
What I got: 
HDR games got washed out colors compared to my windows PC's OLED monitor/tv like LG C2 that got calibrated HDR settings. Red, blue and all other colors look "lifeless". Red looks in games like "orangish" but not red. They don't pop. 
Colors look like in SDR when you would put the vibrancy slider to far left (sRGB). They look washed out.

What I expected:
HDR games showing vibrant colors just like on an OLED monitor connected to a Windows PC.
More info below while telling the steps to reproduce this issue.

#### Steps for reproducing this issue:

1. 
For Native HDR games, start something like Cyberpunk2077, Use HDR Brightness 1000, Midpoint 1.0, Saturation 0 and start a new game. When you are in the bar, look at the Red lights above the pool table. They don't look red compared to my LG C2 OLED monitor/tv. (I have calibrated my C2 and am using the correct HDR settings in-Game). It looks "lifeless" red, more "orang-ish" and washed out. For other colors is the same
(Look at the photo below. It was captured on a iPhone 7 so it looks horrible but still shows the color difference.)
![photo_2023-12-18_18-51-22](https://github.com/ValveSoftware/SteamOS/assets/151800419/f5f33b05-d699-4629-9f0e-728b727329ed)

For other games you can do the same. Go to same location you are on your PC and the Deck OLED, look at colors. FF7R is a great example at beginning when you start. Those blue-ish lights, look washed out on the Deck but not on my OLED monitor.

Another one: Red dead Redemption 2. You start a benchmark with following HDR settings and you will see that your Windows PC + LG C2 for example get normal colors but the colors on the deck OLED are washed out.
HDR mode: Game (not cinematic), Max brightness 1000, Paper White 200 on the Deck OLED. (On your PC's monitor you need to use the correct settings for it. In my case: Max brightness 1000, Paper White 200)

For Ori for example, set base brightness to 0.69 to get almost max 1000nits. Compare that to your OLED HDR capable Device and see how colors look, they look washed out compared to your OLED one. If you put richness at 1.15-1.20 on your Deck, you get "almost" same vibrant colors you got on your OLED monitor.

2. Moonlight/Chiaki have the same issue but the fix for it, is a fix for global HDR settings. What I mean: you make HDR more "vibrant" globaly (for native Deck games) => it makes Moonlight/Chiaki's HDR more vibrant too.

#### How to fix?

Probably by adding an option get more "vibrancy" for HDR. Similar to the SDR vibrancy slider what we got, we could get a vibrancy slider for HDR that makes the colors more saturated. (Not oversaturated but saturated enough that games in HDR don't look washed out)

Thanks in advance.

**Comment by arshiatn** [comment]
I forgot to mention that I am using 75% on HDR Brightness slider of Steam Deck OLED. (Anything above that seems to change the gamma and highlights look overbright). Using "natural color temperature" which can be ticked in display settings in bottom. It is porbably 6500K. Changing the color temperature doesn't fix vibrancy of colors in HDR.

**Comment by roughavoc** [comment]
I want to add some extra information

Testing was done with the DisplayHDR VESA Test App on Windows, streamed via Moonlight

The first Red at 175nits (P3 colourspace), Green (575nits, BT.709 Colourspace) and Blue (60nits, BT.2020)
When compared to an LG OLED TV the Red and Green when shown in P3 and 709 colourspace (from via overlay tools) appear very muted and dull however the Blue, which is BT2020 is just as vivid as the TV

Another slide later on within the app has RGB all displayed in the BT2020 colourspace at 200-300nits, the Steam Deck OLED displays these colours more vividly than the LG OLED

So from this testing, it appears that whatever calibration that was done to the device incorrectly maps Rec709 / DCI-P3 colours

**Comment by KoKlusz** [comment]
https://www.avsforum.com/threads/creators-intent-is-a-little-bit-broken-when-it-comes-to-sdr-gaming.3276018/#post-62577800

This post provides one of the possible explanations why this happens with HDR games.

**Comment by arshiatn** [comment]
> I want to add some extra information
> 
> Testing was done with the DisplayHDR VESA Test App on Windows, streamed via Moonlight (the streaming doesn't change my findings)
> 
> The first of the two slides has a fullscreen Red at (175nits, P3 colourspace), Green (575nits, BT.709 Colourspace) and Blue (60nits, BT.2020) When compared to an LG OLED TV the Red and Green when shown in P3 and 709 colourspace (from via overlay tools) appear very muted and dull however the Blue, which is BT2020 is just as vivid as the TV
> 
> Another slide later on within the app has RGB all displayed in the BT2020 colourspace at 200-300nits, the Steam Deck OLED displays these colours more vividly than the LG OLED
> 
> So from this testing, it appears that whatever calibration that was done to the device incorrectly maps the upper range of Rec709 / DCI-P3 colours

Dude you did so much testing. I really finds my Deck's OLED screen much more vibrant in SDR compared to my LG C2. I think your findings could help a lot.

**Comment by Joshua-Ashton** [comment]
Hello! @arshiatn 

Following this I believe I have identified a bug in the HDR implementation, however, that is making things more saturated, rather than less saturated.

I had a look at Cyberpunk 2077 on my HDR monitor on Windows, and after fixing this bug, they look the same now.

For reference I was using the Steam Deck OLED at 75% brightness for the comparison, so no digital gain was involved.

What is your TV set to, fwir, some LG TVs have a "filmmaker mode", which I think is the one you'll want to use when doing any comparisons like this.

@roughavoc: I believe the Moonlight HDR streaming thing is using libplacebo and doing some tonemapping based on the **host** display (not Deck OLED) right now? Which might be why things look off there.

**Comment by arshiatn** [comment]
@Joshua-Ashton 

My TV was set to "Game Mode" and I was using same color settings of "Hardware Unboxed" on youtube. However after changing to filmmaker mode, which doesn't use HGiG, I put Filmmaker and Game mode to DTM on (Dynamic Tone Mapping on). The both look similar BUT you are right, filmmaker mode is showing a bit more "orang-ish" light above the pool table. It is still not close to Deck's OLED though.

> I had a look at Cyberpunk 2077 on my HDR monitor on Windows, and after fixing this bug, they look the same now.

Basically my monitor is probably not calibrated as I thought. I didn't use any physical hardware to calibrate it :/ so it was a bit expected...
HOWEVER you found a bug in HDR implementation. So, maybe this post helped a bit? xd

1 extra question:

> For reference I was using the Steam Deck OLED at 75% brightness for the comparison, so no digital gain was involved.

So I was seeing it right, that Steam Deck OLED's was playing around with gamma/taking digital gain after 75%, right? 75% seems to giving me 1000nits in brightest objects and still having the most correct HDR compared to anything above 75%.


Thank you VERY much for fixing a HDR Bug. I look forward to a patch/update whenever it drops. I never thought you guys will answer to this issue on github this fast. 
Should I close this issue now?

Edit: about Moonlight, things don't look off on Moonlight for me, they look exactly like how they look on my Steam Deck OLED when I am standing before the pool table in Cyberpun. Same colors and etc. At least to my eyes.

Thanks again and have a great day!

**Comment by pxdl** [comment]
I'm running Steam OS 3.5.7 and still have the same issue. Whenever I run a game with native HDR on my LG CX or LG C1, the colors look way too desaturated, even more so when you compare it to a windows machine plugged to the same TV.

**Comment by roughavoc** [comment]
> I'm running Steam OS 3.5.7 and still have the same issue. Whenever I run a game with native HDR on my LG CX or LG C1, the colors look way too desaturated, even more so when you compare it to a windows machine plugged to the same TV.

@pxdl That is due to an error on the Decks HDR Screen calibration, is fixed in an upcoming Beta / Preview release

**Comment by LordNoxington** [comment]
> > I'm running Steam OS 3.5.7 and still have the same issue. Whenever I run a game with native HDR on my LG CX or LG C1, the colors look way too desaturated, even more so when you compare it to a windows machine plugged to the same TV.
> 
> @pxdl That is due to an error on the Decks HDR Screen calibration, is fixed in an upcoming Beta / Preview release

Is there any evidence of this fix?

**Comment by roughavoc** [comment]
@LordNoxington https://github.com/ValveSoftware/gamescope/commit/52c8cd7e852ff207dd6b6acf875dc53a4677e9c7

**Comment by LordNoxington** [comment]
> @LordNoxington [ValveSoftware/gamescope@52c8cd7](https://github.com/ValveSoftware/gamescope/commit/52c8cd7e852ff207dd6b6acf875dc53a4677e9c7)

Oh amazing. Is there any indication of how frequently repo changes are pushed to preview?

**Comment by Originalimoc** [comment]
HDR streaming from Windows, on 3.5.7 stable, didn't see any problem under my usage, same color with my 1000nit monitor. Might attach a photo later.

**Comment by HerringTheCoder** [comment]
Does the just released SteamOS 3.5.14-preview contain that fix?

**Comment by pxdl** [comment]
> Does the just released SteamOS 3.5.14-preview contain that fix?

I just tested it with the new preview, colors/brightness are still off.

**Comment by LordNoxington** [comment]
> > Does the just released SteamOS 3.5.14-preview contain that fix?
> 
> I just tested it with the new preview, colors/brightness are still off.

Yea can’t see anything in the changelogs about this one. Hope it doesn’t get pushed soon!

**Comment by HerringTheCoder** [comment]
Actually I installed the SteamOS 3.5.14 and there is a huge difference using Moonlight HDR nightly edition by FrogTheFrog. Now the colors are more vibrant than on sdr and even slightly oversaturated(especially reds), but I've messed with my TV HDR settings (including color/saturation) due to the previous issues, so I need to do some further testing. 
SDR content in HDR mode (i.e. browsing the Steam deck library) is finally close to the SDR mode instead of being washed out. Additionally, switching from SDR to HDR does not crash and reset SD anymore.

**Comment by arshiatn** [comment]
> > > Does the just released SteamOS 3.5.14-preview contain that fix?
> > 
> > 
> > I just tested it with the new preview, colors/brightness are still off.
> 
> Yea can’t see anything in the changelogs about this one. Hope it doesn’t get pushed soon!

https://steamcommunity.com/games/1675200/announcements/detail/4064004735511926128?snr=2

I think this one is the fix (3.5.15 Preview). It says: "Improved gamut mapping for HDR titles on Deck OLED".

Can anyone here check it out?

**Comment by roughavoc** [comment]
> Can anyone here check it out?

No change for me

**Comment by Joshua-Ashton** [comment]
Can you give more details as to your issue? We validated Rec.709 Red Green and Blue at different brightness stops here after the change in Preview, so are curious as to are referring to.

**Comment by pxdl** [comment]
> No change for me

Same here. Colors and contrast are still off.

**Comment by Joshua-Ashton** [comment]
Please share an image

**Comment by LordNoxington** [comment]
> Please share an image
SDR:
![IMG_7844](https://github.com/ValveSoftware/SteamOS/assets/89681379/b66ed220-d146-4528-932a-af2dcd69391a)

HDR:
![IMG_7845](https://github.com/ValveSoftware/SteamOS/assets/89681379/f535e835-2202-47b2-9910-7086c196532b)

**Comment by LordNoxington** [comment]
HDR:
![IMG_7845](https://github.com/ValveSoftware/SteamOS/assets/89681379/2990c797-4730-4214-93cf-659b96c84885)

**Comment by roughavoc** [comment]
> Please share an image

Hi all,
Did some more detailed testing

Used an X-Rite i1 DisplayPro - Default Observer bais. Display Type = LED OLED

Measured Rec709 Primaires via Lilliums Shaders within reshade running ontop of Cyberpunk 2077 in both HDR10 and scRGB mode
```
rBt709 = (0.640, 0.330)
gBt709 = (0.300, 0.600)
bBt709 = (0.150, 0.060)
```

Tested at 80nits, 200nits, 400nits, 600nits, 800nits and 1000nits source signal. (Actual output nits is much less due to panel limiations)

I only have access to two HDR displays, those being:

- LG C1 in PC HDMI Input, Game mode, G-Sync off, HGIG on, Colour Gamut = Auto-Detect RGB Full Range - 10 BPC (W11 23H2, NV Driver 551.46)
- Steam Deck OLED - Preview 3.5.15 - Game running via Proton Experimental Bleeding Edge (11/02/24 - 13:00ish)

---

The "issue" is that when displaying 100% r709 primaries instead of the colour only becoming duller/less saturated or just clipping with increase in brightness due to panel limiations, the hue shifts.

Blue becomes Purple/Pink, Red becomes Orange, Green however stays mostly the same.

Now this still happens on the LG OLED but it's very minor. To my eye the difference between 80nits and 800nits is that the image just looks duller despite the brightness increasing.

The IMGSLI below shows measurements at the varying nits, the absolute position of the white dots is not very important, just the relative change between the LG panel and SD panel.

https://imgsli.com/MjM5MzMz/0/1

Something I can't control is the difference between HDR10 and scRGB in the measurements. It could be that the math is slightly off under HDR10 with Liliums shaders, could be that HDR10 just isn't well suited for this testing. But as the same offset was measured on both the LG panel and SD I am going to with HDR10 not being the most accurate for this testing, however it may be something to look into.

Hopefully has been useful.

**Comment by Joshua-Ashton** [comment]
The issue you are seeing with hue shift is not there in our testing. We tested 80, 200, 400, 600, 800 and 1000 nits in our own synthetic test with a Konicka CA-410 and the results were very accurate with no drift.

The issue you are seeing is because your 1000 nit red, is actually trying to achieve a Y (relative luminance) value of 1000 nits with Red, not 1000 nits in the R,G,B sense in which 1,1,1 would be a Y of 1000 nits, but 1,0,0 would be around 210 or so nits.

Thus, the value is being clipped when it is put to the display which is resulting in the hue shift. 

There is potentially some work we can do on our side to prevent hue shifting for values that are clipping (especially for the OS-level HDR gain above 75%), but HDR games should already be tonemapping themselves to the display's luminance range.

Hope this helps!

**Comment by arshiatn** [comment]
> The issue you are seeing with hue shift is not there in our testing. We tested 80, 200, 400, 600, 800 and 1000 nits in our own synthetic test with a Konicka CA-410 and the results were very accurate with no drift.
> 
> The issue you are seeing is because your test for 1000 nit red, is actually trying to achieve a Y (relative luminance) value of 1000 nits with Red, not 1000 nits in the R,G,B sense in which 1,1,1 would be a Y of 1000 nits, but 1,0,0 would be around 210 or so nits.
> 
> Thus, the value is being clipped when it is put to the display which is resulting in the hue shift.
> 
> There is potentially some work we can do on our side to prevent hue shifting for values that are clipping (especially for the OS-level HDR gain above 75%), but HDR games should already be tonemapping themselves to the display's luminance range.
> 
> Hope this helps!

I wrote the following couple of days ago: https://steamcommunity.com/games/1675200/announcements/detail/4064004735511926128?snr=2

I think this one is the fix (3.5.15 Preview). It says: "Improved gamut mapping for HDR titles on Deck OLED".

Can anyone here check it out?

I had time to update to this preview just now and ... it LOOKS AMAZING. Thank you very much. After comparing what I was seeing in Horizon Forbiden West (Chiaki HDR) and Cyberpunk (Native) before the update and now, it is a big upgrade. You were right, I wasn't seeing a "undersaturated" picture but a oversaturated one. Now it looks what I expect a HDR picture to look like. (really similar to my LG C2 now).

Thank you very very much. I hope we see this update in Stable branch soon :)
